runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,129 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_4x8x4_arm_64_dotprod,
     iree_uk_mmt4d_tile_s8s8s32_8x8x4_arm_64_dotprod)
+
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x8x8_to_8x8x8_arm_64_dotprod(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 8 && iree_uk_is_po2_u32(M0));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+  const int8x8_t vzero = vmov_n_s8(0);
+
+  int32x4_t acc[16];
+  for (int i = 0; i < 16; i++) {
+    // We start with zero accumulators and add the value of *out_ptr later.
+    // This is required for the int4 left shift described later.
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  for (int k = 0; k < params->K; ++k) {
+    int8x16_t rhs[4];
+    for (int i = 0; i < 2; i++) {
+      int8x16_t r = vld1q_s8(rhs_ptr);
+      rhs_ptr += 16;
+      // We unpack int4s into individual int8s. To preserve signedness,
+      // int4s are moved to the upper 4-bits of each byte. This has the effect
+      // of multiplying each int4 by 2^4 = 16. To compensate, we divide the
+      // accumulator values by 16 before storing to memory.
+      // This int4 conversion trick is borrowed from the `qd8-f32-qc4w-gemm*`
+      // kernels in https://github.com/google/XNNPACK.
+      rhs[i + 0] = vshlq_n_s8(r, 4);
+      rhs[i + 2] = vandq_s8(r, vmask);
+    }
+
+    int8x8_t lhs[8];
+    if (M0 == 8) {
```

Comment:
This is good to land as-is now. Optionally, if you want to keep on simplifying this code, at this point I think:

- handle the M0==1 and M0==2 cases completely separately (single `if(M0=={1,2})` branch doing both the loads and the dot-products).
- The remaining M0==4 and M0==8 cases can then be merged (M0 being 4 or 8 then merely controls whether a for loop has 1 or 2 iterations) and handle the whole thing (not taking over from the M0==1/M0==2 cases, which makes things unnecessarily complicated).

Note: while the `vdotq_lane_s32` intrinsic wants a `int8x8_t` RHS, and a lane index in the [0,1] range, the underlying SDOT instruction takes 128bit register operands and a lane index in the [0,1,2,3] range. That means that the `vget_low_s8` intrinsics get fused with the `vdotq_lane_s32` into a single SDOT instruction. That's why this kernel needs fewer registers than it may look like, and why writing the vget_low/high_s8 in the same statement as the vdotq_lane_s32 is actually idiomatic in terms of what this compiles to.

Reply:
I changed the code to handle M0={1,2} and M0={4,8} loops separately.

> Note: while the vdotq_lane_s32 intrinsic wants a int8x8_t RHS, and a lane index in the [0,1] range, the underlying SDOT instruction takes 128bit register operands and a lane index in the [0,1,2,3] range. That means that the vget_low_s8 intrinsics get fused with the vdotq_lane_s32 into a single SDOT instruction. That's why this kernel needs fewer registers than it may look like, and why writing the vget_low/high_s8 in the same statement as the vdotq_lane_s32 is actually idiomatic in terms of what this compiles to.

Thanks for the info. I initially separated them because the previous ukernel PR showed slowdowns when using vget_low/high but make sense to look at the underlying instruction. I'll leave it as is since changing it to use vget* will modify the loop structure.